ORA-01400 crash when archiving a conversation that involves a bare JID (no resource) on Oracle

As reported in https://discourse.igniterealtime.org/t/error-sending-message/93852

When Openfire is backed by an Oracle database and the Monitoring plugin is installed, storing chat message can trigger the following error in the Openfire log, and the conversation metadata is not persisted.

Oracle's SQL standard deviation causes this. Oracle treats an empty string (`''`) as `NULL`. The `ofConParticipant.jidResource` column is defined `NOT NULL` in the schema. When a message is sent from a bare JID (one without a resource, as is common for server-to-client messages), the code passed an empty string for `jidResource`. On all databases other than Oracle this is stored verbatim; on Oracle it is silently coerced to `NULL`, violating the constraint.

```
14:58:44.740 [socket_c2s-thread-4] ERROR org.jivesoftware.openfire.archive.ConversationDAO - Unable to persist a conversation that was just created: Conversation [11] between {xxxxx=org.jivesoftware.openfire.archive.UserParticipations@1bcc99ea, admin@xxxxxxx=org.jivesoftware.openfire.archive.UserParticipations@1bcc99ea}. started Apr 16, 2024, 2:58:44â€¯PM, last active Apr 16, 2024, 2:58:44â€¯PM. Total messages: 0
java.sql.SQLIntegrityConstraintViolationException: ORA-01400: impossibile inserire NULL in ("OPENFIRE"."OFCONPARTICIPANT"."JIDRESOURCE")

https://docs.oracle.com/error-help/db/ora-01400/
	at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:702) ~[ojdbc11.jar:23.3.0.23.09]
	at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:608) ~[ojdbc11.jar:23.3.0.23.09]
	at oracle.jdbc.driver.T4C8Oall.processError(T4C8Oall.java:1330) ~[ojdbc11.jar:23.3.0.23.09]
	at oracle.jdbc.driver.T4CTTIfun.receive(T4CTTIfun.java:1102) ~[ojdbc11.jar:23.3.0.23.09]
	at oracle.jdbc.driver.T4CTTIfun.doRPC(T4CTTIfun.java:456) ~[ojdbc11.jar:23.3.0.23.09]
	at oracle.jdbc.driver.T4C8Oall.doOALL(T4C8Oall.java:498) ~[ojdbc11.jar:23.3.0.23.09]
	at oracle.jdbc.driver.T4CPreparedStatement.doOall8(T4CPreparedStatement.java:272) ~[ojdbc11.jar:23.3.0.23.09]
	at oracle.jdbc.driver.T4CPreparedStatement.executeForRows(T4CPreparedStatement.java:1451) ~[ojdbc11.jar:23.3.0.23.09]
	at oracle.jdbc.driver.OracleStatement.executeSQLStatement(OracleStatement.java:1958) ~[ojdbc11.jar:23.3.0.23.09]
	at oracle.jdbc.driver.OracleStatement.doExecuteWithTimeout(OracleStatement.java:1594) ~[ojdbc11.jar:23.3.0.23.09]
	at oracle.jdbc.driver.OraclePreparedStatement.executeInternal(OraclePreparedStatement.java:3754) ~[ojdbc11.jar:23.3.0.23.09]
	at oracle.jdbc.driver.OraclePreparedStatement.executeLargeUpdate(OraclePreparedStatement.java:4100) ~[ojdbc11.jar:23.3.0.23.09]
	at oracle.jdbc.driver.OraclePreparedStatement.executeUpdate(OraclePreparedStatement.java:4071) ~[ojdbc11.jar:23.3.0.23.09]
	at oracle.jdbc.driver.OraclePreparedStatementWrapper.executeUpdate(OraclePreparedStatementWrapper.java:997) ~[ojdbc11.jar:23.3.0.23.09]
	at org.apache.commons.dbcp2.DelegatingPreparedStatement.executeUpdate(DelegatingPreparedStatement.java:136) ~[commons-dbcp2-2.9.0.jar:2.9.0]
	at org.apache.commons.dbcp2.DelegatingPreparedStatement.executeUpdate(DelegatingPreparedStatement.java:136) ~[commons-dbcp2-2.9.0.jar:2.9.0]
	at org.jivesoftware.openfire.archive.ConversationDAO.insertIntoDb(ConversationDAO.java:321) ~[?:?]
	at org.jivesoftware.openfire.archive.ConversationDAO.createConversation(ConversationDAO.java:88) ~[?:?]
	at org.jivesoftware.openfire.archive.ConversationManager.processMessage(ConversationManager.java:751) ~[?:?]
	at org.jivesoftware.openfire.archive.ArchiveInterceptor.interceptPacket(ArchiveInterceptor.java:86) ~[?:?]
	at org.jivesoftware.openfire.interceptor.InterceptorManager.invokeInterceptors(InterceptorManager.java:268) ~[xmppserver-4.8.1.jar:4.8.1]
	at org.jivesoftware.openfire.interceptor.InterceptorManager.invokeInterceptors(InterceptorManager.java:230) ~[xmppserver-4.8.1.jar:4.8.1]
	at org.jivesoftware.openfire.MessageRouter.route(MessageRouter.java:170) ~[xmppserver-4.8.1.jar:4.8.1]
	at org.jivesoftware.openfire.spi.PacketRouterImpl.route(PacketRouterImpl.java:79) ~[xmppserver-4.8.1.jar:4.8.1]
	at org.jivesoftware.openfire.net.StanzaHandler.processMessage(StanzaHandler.java:434) ~[xmppserver-4.8.1.jar:4.8.1]
	at org.jivesoftware.openfire.net.ClientStanzaHandler.processMessage(ClientStanzaHandler.java:104) ~[xmppserver-4.8.1.jar:4.8.1]
	at org.jivesoftware.openfire.net.StanzaHandler.process(StanzaHandler.java:258) ~[xmppserver-4.8.1.jar:4.8.1]
	at org.jivesoftware.openfire.net.StanzaHandler.processStanza(StanzaHandler.java:222) ~[xmppserver-4.8.1.jar:4.8.1]
	at org.jivesoftware.openfire.net.StanzaHandler.process(StanzaHandler.java:114) ~[xmppserver-4.8.1.jar:4.8.1]
	at org.jivesoftware.openfire.nio.NettyConnectionHandler.channelRead0(NettyConnectionHandler.java:142) ~[xmppserver-4.8.1.jar:4.8.1]
	at org.jivesoftware.openfire.nio.NettyConnectionHandler.channelRead0(NettyConnectionHandler.java:50) ~[xmppserver-4.8.1.jar:4.8.1]
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346) ~[netty-codec-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318) ~[netty-codec-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286) ~[netty-handler-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.handler.traffic.AbstractTrafficShapingHandler.channelRead(AbstractTrafficShapingHandler.java:506) ~[netty-handler-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1471) ~[netty-handler-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1334) ~[netty-handler-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1383) ~[netty-handler-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:529) ~[netty-codec-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:468) ~[netty-codec-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:290) ~[netty-codec-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562) ~[netty-transport-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997) ~[netty-common-4.1.100.Final.jar:4.1.100.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.100.Final.jar:4.1.100.Final]
	at java.lang.Thread.run(Thread.java:1570) [?:?]
Caused by: oracle.jdbc.OracleDatabaseException: ORA-01400: impossibile inserire NULL in ("OPENFIRE"."OFCONPARTICIPANT"."JIDRESOURCE")

	at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:710) ~[ojdbc11.jar:23.3.0.23.09]
	... 68 more
```

--- a/changelog.html
+++ b/changelog.html
@@ -47,6 +47,7 @@ Monitoring Plugin Changelog
 <p><b>2.8.0</b> -- (to be determined)</p>
 <ul>
     <li>Requires Openfire 5.1.0</li>
+    <li>[<a href='https://github.com/igniterealtime/openfire-monitoring-plugin/issues/461'>Issue #461</a>] - ORA-01400 crash when archiving a conversation that involves a bare JID (no resource) on Oracle</li>
     <li>[<a href='https://github.com/igniterealtime/openfire-monitoring-plugin/issues/441'>Issue #441</a>] - Enable message archiving by default</li>
     <li>[<a href='https://github.com/igniterealtime/openfire-monitoring-plugin/issues/439'>Issue #439</a>] - Prevent leaking real JIDs through message archive when querying MUC private messages</li>
     <li>[<a href='https://github.com/igniterealtime/openfire-monitoring-plugin/issues/436'>Issue #436</a>] - Prevent database storage issue when using statistics with a key that's longer than 20 characters</li>

--- a/plugin.xml
+++ b/plugin.xml
@@ -6,7 +6,7 @@
     <description>Monitors conversations and statistics of the server.</description>
     <author>Ignite Realtime</author>
     <version>${project.version}</version>
-    <date>2025-12-09</date>
+    <date>2026-04-06</date>
     <minServerVersion>5.1.0</minServerVersion>
     <databaseKey>monitoring</databaseKey>
     <databaseVersion>10</databaseVersion>

--- a/src/java/org/jivesoftware/openfire/archive/ConversationDAO.java
+++ b/src/java/org/jivesoftware/openfire/archive/ConversationDAO.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2021-2025 Ignite Realtime Foundation. All rights reserved.
+ * Copyright (C) 2021-2026 Ignite Realtime Foundation. All rights reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -374,7 +374,7 @@ public class ConversationDAO {
                 // Rebuild full JID of participant
                 String baredJID = rs.getString(1);
                 String resource = rs.getString(2);
-                JID fullJID = new JID("".equals(resource) ? baredJID : baredJID + "/" + resource);
+                JID fullJID = new JID(hasStoredResource(resource) ? baredJID + "/" + resource : baredJID);
                 // Rebuild joined and left time
                 ConversationParticipation participation = new ConversationParticipation(new Date(rs.getLong(4)), rs.getString(3));
                 if (rs.getLong(5) > 0) {
@@ -432,7 +432,7 @@ public class ConversationDAO {
                     pstmt.setLong(1, conversation.getConversationID());
                     pstmt.setLong(2, participation.getJoined().getTime());
                     pstmt.setString(3, user.toBareJID());
-                    pstmt.setString(4, user.getResource() == null ? "" : user.getResource());
+                    pstmt.setString(4, toStoredResource(user));
                     pstmt.setString(5, participation.getNickname());
                     pstmt.executeUpdate();
                 }
@@ -466,12 +466,48 @@ public class ConversationDAO {
             pstmt.setLong(1, conversationID);
             pstmt.setLong(2, joined);
             pstmt.setString(3, participant.toBareJID());
-            pstmt.setString(4, participant.getResource() == null ? "" : participant.getResource());
+            pstmt.setString(4, toStoredResource(participant));
             pstmt.setString(5, nickname);
             pstmt.executeUpdate();
             pstmt.close();
         } finally {
             DbConnectionManager.closeConnection(con);
         }
+    }
+
+    /**
+     * Converts a JID resource into the value that is stored in the {@code jidResource} column of {@code ofConParticipant}.
+     *
+     * Oracle treats empty strings as {@code NULL}, which would violate the {@code NOT NULL} constraint on
+     * {@code jidResource}. When the database is Oracle and the JID has no resource (i.e. it is a bare JID),
+     * a single-space sentinel value is used instead of an empty string. On all other databases the historic
+     * empty-string value is preserved.
+     *
+     * @param jid the JID whose resource part is to be stored.
+     * @return the value to store in the {@code jidResource} column; never {@code null}.
+     * @see <a href="https://github.com/igniterealtime/openfire-monitoring-plugin/issues/461">ORA-01400 crash when archiving a conversation that involves a bare JID (no resource) on Oracle</a>
+     */
+    static String toStoredResource(final JID jid)
+    {
+        if (jid.getResource() != null) {
+            return jid.getResource();
+        }
+        return DbConnectionManager.getDatabaseType() == DbConnectionManager.DatabaseType.oracle ? " " : "";
+    }
+
+    /**
+     * Determines whether a value read from the {@code jidResource} column of {@code ofConParticipant} represents
+     * a real XMPP resource.
+     *
+     * Both an empty string (used by non-Oracle databases for bare JIDs) and the single-space sentinel (used by
+     * Oracle, because Oracle treats empty strings as {@code NULL}) are considered "no resource" values and cause
+     * this method to return {@code false}.
+     *
+     * @param storedResource the raw value read from the database column.
+     * @return {@code true} if the value represents an actual JID resource; {@code false} if it is absent or a sentinel.
+     */
+    static boolean hasStoredResource(final String storedResource)
+    {
+        return storedResource != null && !storedResource.trim().isEmpty();
     }
 }

--- a/src/java/org/jivesoftware/openfire/archive/ConversationManager.java
+++ b/src/java/org/jivesoftware/openfire/archive/ConversationManager.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2008 Jive Software, Ignite Realtime Foundation 2024-2025. All rights reserved.
+ * Copyright (C) 2008 Jive Software, Ignite Realtime Foundation 2024-2026. All rights reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -1432,7 +1432,7 @@ public class ConversationManager implements ComponentEventListener{
                     pstmt.setLong(1, work.left.getTime());
                     pstmt.setLong(2, work.conversationID);
                     pstmt.setString(3, work.user.toBareJID());
-                    pstmt.setString(4, work.user.getResource() == null ? " " : work.user.getResource());
+                    pstmt.setString(4, ConversationDAO.toStoredResource(work.user));
                     pstmt.setLong(5, work.joined.getTime());
                     if ( DbConnectionManager.isBatchUpdatesSupported() )
                     {

